p2p/net/swarm/black_hole_detector.go

```diff
@@ -0,0 +1,257 @@
+	blackHoleStateAllowed
+	blackHoleStateBlocked
+)
+
+func (st blackHoleState) String() string {
+	switch st {
+	case blackHoleStateProbing:
+		return "Probing"
+	case blackHoleStateAllowed:
+		return "Allowed"
+	case blackHoleStateBlocked:
+		return "Blocked"
+	default:
+		return fmt.Sprintf("Unknown %d", st)
+	}
+}
+
+type blackHoleResult int
+
+const (
+	blackHoleResultAllowed blackHoleResult = iota
+	blackHoleResultProbing
+	blackHoleResultBlocked
+)
+
+// blackHoleFilter provides black hole filtering for dials. This filter should be used in
+// concert with a UDP of IPv6 address filter to detect UDP or IPv6 black hole. In a black
+// holed environments dial requests are blocked and only periodic probes to check the
+// state of the black hole are allowed.
+//
+// Requests are blocked if the number of successes in the last n dials is less than
+// minSuccesses. If a request succeeds in Blocked state, the filter state is reset and n
+// subsequent requests are allowed before reevaluating black hole state. Dials cancelled
+// when some other concurrent dial succeeded are counted as failures. A sufficiently large
+// n prevents false negatives in such cases.
+type blackHoleFilter struct {
+	// n serves the dual purpose of being the minimum number of requests after which we
+	// probe the state of the black hole in blocked state and the minimum number of
+	// completed dials required before evaluating black hole state.
+	n int
+	// minSuccesses is the minimum number of Success required in the last n dials
+	// to consider we are not blocked.
+	minSuccesses int
+	// name for the detector.
+	name string
+
+	// requests counts number of dial requests to peers. We handle request at a peer
+	// level and record results at individual address dial level.
+	requests int
+	// dialResults of the last `n` dials. A successful dial is true.
+	dialResults []bool
+	// successes is the count of successful dials in outcomes
+	successes int
+	// state is the current state of the detector
+	state blackHoleState
+
+	mu            sync.Mutex
+	metricsTracer MetricsTracer
+}
+
+// RecordResult records the outcome of a dial. A successful dial will change the state
+// of the filter to Allowed. A failed dial only blocks subsequent requests if the success
+// fraction over the last n outcomes is less than the minSuccessFraction of the filter.
+func (b *blackHoleFilter) RecordResult(success bool) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if b.state == blackHoleStateBlocked && success {
+		// If the call succeeds in a blocked state we reset to allowed.
+		// This is better than slowly accumulating values till we cross the minSuccessFraction
+		// threshold since a blackhole is a binary property.
+		b.reset()
+		return
+	}
+
+	if success {
+		b.successes++
+	}
+	b.dialResults = append(b.dialResults, success)
+
+	if len(b.dialResults) > b.n {
+		if b.dialResults[0] {
+			b.successes--
+		}
+		b.dialResults = b.dialResults[1 : b.n+1]
```

Comment:
This is equivalent to
```suggestion
		b.dialResults = b.dialResults[1:]
```
right?